doc/sphinx/src/using-eos.rst

```diff
@@ -39,6 +40,271 @@ A more generic version of the vector calls exists in the ``Evaluate``
+have a single pointer containing all serialized EOS's. Same for the
+shared memory. ``singularity-eos`` provides machinery to do this in
+the ``singularity-eos/base/serialization_utils.hpp`` header. This
+provides a helper struct, ``BulkSerializer``:
+
+.. code-block:: cpp
+
+  template<typename Container_t, Resizer_t = MemberResizer>
+  singularity::BulkSerializer
+
+which may be initialized by a collection of ``EOS`` objects or by
+simply assigning (or constructing) its member field, ``eos_objects``
+appropriately. An example ``Container_t`` might be
+``std::vector<EOS>``. A specialization for ``vector`` is provided as
+``VectorSerializer``. The ``Resizer_t`` is a functor that knows how to
+resize a collection. For example, the ``MemberResizor`` functor used
+for ``std::vector``s
+
+.. code-block:: cpp
+
+  struct MemberResizer {
+    template<typename Collection_t>
+    void operator()(Collection_t &collection, std::size_t count) {
+      collection.resize(count);
+    }
+  };
+
+which will work for any ``stl`` container with a ``resize`` method.
+
+The ``BulkSerializer`` provides all the above-described serialization
+functions for ``EOS`` objects: ``SerializedSizeInBytes``,
+``SharedMemorySizeInBytes``, ``Serialize``, and ``DeSerialize``, but
+it operates on all ``EOS`` objects contained in the container it
+wraps, not just one. Example usage might look like this:
+
+.. code-block:: cpp
+
+  int packed_size, shared_size;
+  singularity::VectorSerializer<EOS> serializer;
+  if (rank == 0) { // load eos object
+    // Code to initialize a bunch of EOS objects into a std::vector<EOS>
+    /*
+       Initialization code goes here
+     */
+    serializer = singularity::VectorSerializer<EOS>(eos_vec);
+    packed_size = serializer.SerializedSizeInBytes();
+    shared_size = serializer.SharedMemorySizeInBytes();
+  }
+
+  // Send sizes
+  MPI_Bcast(&packed_size, 1, MPI_INT, 0, MPI_COMM_WORLD);
+  MPI_Bcast(&packed_size, 1, MPI_INT, 0, MPI_COMM_WORLD);
+
+  // Allocate data needed
+  packed_data = (char*)malloc(packed_size);
+  if (rank == 0) {
+    serializer.Serialize(packed_data);
+    serializer.Finalize(); // Clean up all EOSs owned by the serializer
+  }
+  MPI_Bcast(packed_data, packed_size, MPI_BYTE, 0, MPI_COMM_WORLD);
+  
+  singularity::SharedMemSettings settings = singularity::DEFAULT_SHMEM_SETTINGS;
+  // same MPI declarations as above
+  if (use_mpi_shared_memory) {
+    // same MPI code as above including setting the settings
+    settings.data = shared_data;
+    settings.is_domain_root = (island_rank == 0);
+  }
+  singularity::VectorSerializer<EOS> deserializer;
+  deserializer.DeSerialize(packed_data, settings);
+  if (use_mpi_shared_memory) {
+    // same MPI code as above
+  }
+  // extract each individual EOS and do something with it
+  std::vector<EOS> eos_host_vec = deserializer.eos_objects;
+  // get on device if you want
+  for (auto EOS : eos_host_vec) {
```

Comment:
This could be a can of worms at some point when we want to have 1 copy per GPU rather than 1 copy per node. Some host codes may want to hand us GPU memory so we may need to have a facility to handle that case at some point.

Reply:
What this will *currently* produce is 1 copy per MPI rank on device and 1 copy per node on host, because `GetOnDevice` will do its own allocation. The case where we want device shared memory is one I'm aware of but haven't thought deeply about. My suggestion is we punt on this but agree there's an engineering problem to solve here.